DEFEDIT-1337 Reduce memory usage during build

erikangelin-king opened the pull request:
* User facing changes
  - Less memory used for json (spine) resources
  - Less memory used/cached during build
  - Disk cache more often used when building

* Technical changes
  - JsonNodes no longer store clojurified json data in property,
  instead parsed on demand as part of output transform
  - build-target outputs for several types no longer produce data up
  front and put it in user-data (and thus also graph cache) for
  build-fn's to use. Instead they pass along "generators" - functions
  + arguments for producing the needed data. This also avoids
  non-value types in user-data, as those break the disk artifact cache
  lookup.
  - generate-xxx functions produce a value or g/error
  - produce-xxx-generator attempts to validate arguments as far as reasonable
  - build-targets outputs should mostly collect arguments/data from
  graph and perform cheap validation. Meaty work/validation (like
  reading external resources) should be performed in build-xxx
  - build-xxx should now produce a build artifact as before or a
  node-id-less g/error
  - pipeline/build! now needs to be prepared for g/errors from build
  functions and will decorate these with the corresponding _node-id
  for later error reporting
  - Renamed several functions, project/build! project/build-project!

erikangelin-king committed:
DEFEDIT-1337 Reduce memory usage during build

* User facing changes
  - Less memory used for json (spine) resources
  - Less memory used/cached during build
  - Disk cache more often used when building

* Technical changes
  - JsonNodes no longer store clojurified json data in property,
  instead parsed on demand as part of output transform
  - build-target outputs for several types no longer produce data up
  front and put it in user-data (and thus also graph cache) for
  build-fn's to use. Instead they pass along "generators" - functions
  + arguments for producing the needed data. This also avoids
  non-value types in user-data, as those break the disk artifact cache
  lookup.
  - generate-xxx functions produce a value or g/error
  - produce-xxx-generator attempts to validate arguments as far as reasonable
  - build-targets outputs should mostly collect arguments/data from
  graph and perform cheap validation. Meaty work/validation (like
  reading external resources) should be performed in build-xxx
  - build-xxx should now produce a build artifact as before or a
  node-id-less g/error
  - pipeline/build! now needs to be prepared for g/errors from build
  functions and will decorate these with the corresponding _node-id
  for later error reporting
  - Renamed several functions, project/build! project/build-project!

matgis-king reviewed (changes requested):
Partial review up to `texture_set_gen.clj`. Some concerns.

erikangelin-king committed:
DEFEDIT-1337 Review fixes

* atlas: scan for g/errors in shas, not concatted sha string
* image: image-util/read-image + read-size throw exception on failure,
  translated by validation/resource-io-with-errrors to an
  invalid-content-error.
* json: use PushbackReader when reading instead of slurping to string
* pipeline: weird whitespace
* rig: change map -> (pb-message -> ) pb string -> hash to map ->
  pb-message -> hash
* tile-source: whitespace

erikangelin-king committed:
DEFEDIT-1337 Review fixes

Build now displays "Compiling ..." followed by "Building ..."
